easybuild/tools/module_generator.py

```diff
@@ -257,11 +246,38 @@ def prepend_paths(self, key, paths, allow_abs=False, expand_relpaths=True):
         :param allow_abs: allow providing of absolute paths
         :param expand_relpaths: expand relative paths into absolute paths (by prefixing install dir)
         """
-        paths = self._filter_paths(key, paths)
-        if paths is None:
-            return ''
         return self.update_paths(key, paths, prepend=True, allow_abs=allow_abs, expand_relpaths=expand_relpaths)
 
+    def update_paths(self, key, paths, prepend=True, allow_abs=False, expand_relpaths=True):
+        """
+        Generate append/prepend-path statements for the given list of paths.
+
+        :param key: environment variable to append paths to
+        :param paths: list of paths to append
+        :param allow_abs: allow providing of absolute paths
+        :param expand_relpaths: expand relative paths into absolute paths (by prefixing install dir)
+        """
+        if isinstance(paths, str):
+            self.log.debug("Wrapping %s into a list before using it for %s", paths, key)
+            paths = [paths]
+
+        if key == 'PYTHONPATH':
+            python_paths = [path for path in paths if re.match(r'lib/python\d+\.\d+/site-packages', path)]
```

Comment:
Checking the logic in `_filter_paths` I see an issue here: You iterate `paths` multiple times: Here and at https://github.com/easybuilders/easybuild-framework/pull/4496/files#diff-4af72e9777d353325a29df400ab4229e14defd653e413db51786dc2172d3baefR275 but inside `_update_paths` you call `_filter_paths` which "coerces a generator into a list" if it isn't a list yet.

This means it this point `paths` might be a generator which can only be iterated once as otherwise the code in `_filter_paths` wasn't required.

Similar to my previous reasoning about breaking change in favor of cleaner code: I would simply disallow passing anything but a a string or a list to `update_paths` and assert that it is a list in `_filter_paths` instead of trying to convert a potentially exhausted generator to a list here or there.

Quick example:
```
paths=(i for i in range(10))
[i for i in paths if i==2] # [2]
[i for i in paths if i==2] # []
```